src/hotspot/share/services/heapDumper.cpp

```diff
@@ -787,6 +789,111 @@ class DumperSupport : AllStatic {
   }
 };
 
+// Hash table of klasses to the klass metadata. This should greatly improve the
+// hash dumping performance. This hash table is supposed to be used by a single
+// thread only.
+//
+class DumperClassCacheTableEntry : public CHeapObj<mtServiceability> {
+  friend class DumperClassCacheTable;
+private:
+  GrowableArray<char> _sigs_start;
+  GrowableArray<int> _offsets;
+  u4 _instance_size;
+  int _entries;
+
+public:
+  DumperClassCacheTableEntry() : _instance_size(0), _entries(0) {};
+
+  int field_count()             { return _entries; }
+  char sig_start(int field_idx) { return _sigs_start.at(field_idx); }
+  int offset(int field_idx)     { return _offsets.at(field_idx); }
+  u4 instance_size()            { return _instance_size; }
+};
+
+class DumperClassCacheTable : public StackObj {
+private:
+  // ResourceHashtable SIZE is specified at compile time so we
+  // use 1031 which is the first prime after 1024.
+  static constexpr size_t TABLE_SIZE = 1031;
+
+  // Maintain the cache for N classes. This limits memory footprint
+  // impact, regardless of how many classes we have in the dump.
+  // This also improves look up performance by keeping the statically
+  // sized table from overloading.
+  static constexpr int CACHE_TOP = 256;
+
+  static unsigned int ptr_hash(void* const& s1) {
```

Comment:
Maybe? Need to brush up my knowledge on non-power-of-two hash tables. This thing just copies the hash we have in other places :) But since we are looking at it, I see no grand point in overriding the hash code here, and we can just use the default one. When/if default one improves, this one should improve as well.

Reply:
The reason I asked is that we use this primitive_hash for a lot of hashtables of InstanceKlass and Symbol so if your experiments saw too many collisions, it would be good to change for everything.

Reply:
No, I just copy-pasted it without thinking all too much about the existence of the default hash code. I don't see much, if any, difference with default hashcode in my quick tests. So I left the default.